File: python/1-D Dynamic Programming/palindromic_substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        # res: stores the total number of palindromic substrings
        res = 0
        
        for i in range(len(s)):
            # Check for odd-length palindromes centered at i
            l = r = i
            while l >= 0 and r < len(s) and s[l] == s[r]:
                res += 1
                l -= 1
                r += 1

            # Check for even-length palindromes centered between i and i+1
            l, r = i, i + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                res += 1
                l -= 1
                r += 1
        return res

# Time Complexity: O(n^2), where n is the length of the string. For each character, we expand around the center up to n times.
# Space Complexity: O(1), since only a constant amount of extra space is used.

# Better modularity of code
class Solution2:
    def countSubstrings(self, s: str) -> int:
        # res: stores the total number of palindromic substrings
        res = 0
        
        for i in range(len(s)):
            # Count odd-length palindromes centered at i
            res += self.countPali(s, i, i)
            # Count even-length palindromes centered between i and i+1
            res += self.countPali(s, i, i + 1)
        
        return res
        
    def countPali(self, s, l, r):
        # Helper function to count palindromes expanding from the center (l, r)
        res = 0
        while l >= 0 and r < len(s) and s[l] == s[r]:
            res += 1
            l -= 1
            r += 1
        return res

# Time Complexity: O(n^2), where n is the length of the string. For each character, we expand around the center up to n times.
# Space Complexity: O(1), since only a constant amount of extra space is used.

# -------------------
# Test Cases
# -------------------
def test_palindromic_substrings():
    s1 = Solution()
    s2 = Solution2()
    # Test case 1: Example from Leetcode
    assert s1.countSubstrings("abc") == 3
    assert s2.countSubstrings("abc") == 3
    # Test case 2: Another example
    assert s1.countSubstrings("aaa") == 6
    assert s2.countSubstrings("aaa") == 6
    # Test case 3: Single character
    assert s1.countSubstrings("a") == 1
    assert s2.countSubstrings("a") == 1
    # Test case 4: All characters different
    assert s1.countSubstrings("abcd") == 4
    assert s2.countSubstrings("abcd") == 4
    # Test case 5: Empty string
    assert s1.countSubstrings("") == 0
    assert s2.countSubstrings("") == 0
    print("All test cases passed!")

if __name__ == "__main__":
    test_palindromic_substrings()

File: C++/1-D Dynamic Programming/palindromic_substrings.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int countSubstrings(string s) {
        /*
            Problem:
            - Count all palindromic substrings in a given string.

            Approach:
            - Expand Around Center:
                - For each character `i` in the string:
                    - Treat it as the center of an odd-length palindrome and expand.
                    - Treat it as the left of center of an even-length palindrome and expand.
                - Increment the count for every valid palindrome found.

            Time Complexity: O(n^2)
                - For each character, expand up to n times.
            Space Complexity: O(1)
                - Only variables for indices and count.
        */

        int res = 0; // Total number of palindromic substrings

        for (int i = 0; i < s.size(); ++i) {
            // Odd-length palindromes centered at i
            int l = i, r = i;
            while (l >= 0 && r < s.size() && s[l] == s[r]) {
                res++;
                l--;
                r++;
            }

            // Even-length palindromes centered between i and i+1
            l = i, r = i + 1;
            while (l >= 0 && r < s.size() && s[l] == s[r]) {
                res++;
                l--;
                r++;
            }
        }

        return res;
    }
};

/*
================ LINE-BY-LINE SYNTAX EXPLANATION =================

#include <bits/stdc++.h>
    - Includes all standard C++ libraries.

using namespace std;
    - Allows usage of standard library classes and functions without `std::` prefix.

class Solution {
    - Defines a class named Solution.

public:
    - Makes the methods below publicly accessible.

int countSubstrings(string s)
    - Function `countSubstrings` takes a string `s` and returns an int.

int res = 0;
    - Initialize a counter to store the number of palindromic substrings.

for (int i = 0; i < s.size(); ++i)
    - Loop over each character in the string.

int l = i, r = i;
    - Initialize two pointers at the same index to check odd-length palindromes.

while (l >= 0 && r < s.size() && s[l] == s[r])
    - Expand outward while the characters at `l` and `r` are equal.

res++;
    - Increment count for the found palindrome.

l--, r++;
    - Move pointers outward to check for longer palindromes.

l = i, r = i + 1;
    - Initialize pointers for even-length palindrome check.

while (l >= 0 && r < s.size() && s[l] == s[r])
    - Same as above but for even-length palindromes.

return res;
    - Return the total number of palindromic substrings found.

*/
